---
title: "Genomic Region Profiles"
author: "Stephen Kelly"
date: "1/17/2017"
output: 
  html_document: 
    keep_md: yes
    toc: true
    toc_float: true
    toc_depth: 3    
    number_sections: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup & Workflow {.tabset .tabset-pills}

## Hide

## Show

### Find BigWigs

First, start a sample sheet with the paths to all of the bigwig files we will need to process. 

```{r, engine='bash', eval=FALSE}
# module unload python
# deeptools/2.3.3 # computeMatrix 1.5.11


project_dir="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10"
region_dir="${project_dir}/project_notes/region_profiles"
cd "$region_dir"

# find the bigwigs from the pipeline, make a sample sheet
find "${project_dir}/pipeline/align/" -type f -name "track.bw" -exec readlink -f {} \; > bigwig_sample_sheet_og.tsv

# manually edit the sample sheet in Excel
```

Set up a directory with symlinks to all of the bigwig files, for easy access. 

```{r, engine='bash', eval=FALSE}
# set up the bigwig links
project_dir="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10"
region_dir="${project_dir}/project_notes/region_profiles"
cd "$region_dir"

bigwig_dir="${region_dir}/bigwigs"
matrix_dir="${region_dir}/matricies"

mkdir -p "$bigwig_dir"
mkdir -p "$matrix_dir"

sample_sheet="${region_dir}/bigwig_sample_sheet.tsv"

# iterate over entries in the sample sheet
tail -n +2 "$sample_sheet"| while read line; do
if [ ! -z "$line" ]; then
filepath="$(echo "$line" | cut -f1)"

link_name="$(echo "$line" | cut -f5).bw"

echo "$filepath"
echo "$link_name"
echo ""

(cd "$bigwig_dir"
ln -fs "$filepath" "$link_name"
)


fi
done
```

### Install deepTools

We will need to install deepTools. A Python virtual environment will be set up for this in order to isolate it from the rest of the installed system packages. 

```{r, engine='bash', eval=FALSE}
# set up DeepTools
module unload python
unset PYTHONPATH

project_dir="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10"
cd "$project_dir"

virtualenv venv --no-site-packages
source venv/bin/activate
pip install deeptools

# check the install
plotProfile --version # plotProfile 2.4.2
computeMatrix --version # computeMatrix 2.4.2

# http://deeptools.readthedocs.io/en/latest/content/tools/plotProfile.html
# http://deeptools.readthedocs.io/en/latest/content/tools/computeMatrix.html
```

### Get Gencode Genes

A set of gene locations will be obtained from Gencode, and the GTF file will be converted to BED format with BEDOPS, and subset for only gene entries. 

```{r, engine='bash', eval=FALSE}
# get the Gencode genes file to use
module unload gcc
module load bedtools/2.22.0

# file prefix
gen_file="gencode.v19.annotation"
# Download source data ; http://www.gencodegenes.org/releases/19.html
[ ! -f ${gen_file}.gtf.gz ] && wget "ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_19/gencode.v19.annotation.gtf.gz"

# convert the GTF to BED witht he BEDOPS converter
[ ! -f ${gen_file}.bed ] && zcat ${gen_file}.gtf.gz | convert2bed --input=gtf - > ${gen_file}.bed


# get only the 'gene' entries from the BED file
grep -E '[[:space:]]gene[[:space:]]' ${gen_file}.bed > ${gen_file}.genes_only.bed
```

### Get DiffBind Data

The DiffBind ChIP-Seq datasets for differentially bound regions between samples per histone marks will be found and linked


```{r, engine='bash', eval=FALSE}
project_dir="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10"
region_dir="${project_dir}/project_notes/region_profiles"
cd "$region_dir"

diffbind_dir="${region_dir}/diffbind_sheets"
mkdir -p "$diffbind_dir"

diffbind_data_location="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10/project_notes/results_dir/diffbind_DESeq2/"

diffbind_files="$(find "$diffbind_data_location" -type f -name "diff_bind.*" -name "*.csv" -exec readlink -f {} \;)"

for diff_file in $diffbind_files; do
(
echo "$diff_file"
hist_mark="$(basename "$(dirname "$diff_file")")"
echo "$hist_mark"
out_name="$(echo "$(basename "$diff_file")" | sed "s/diff_bind/$hist_mark/")"
echo "$out_name"

cd "$diffbind_dir"
ln -fs "$diff_file" "$out_name"
)
done


```

### Match BigWigs to DiffBind sheets

A sample sheet will be made to match each BigWig sample file with the corresponding DiffBind sheet

```{r, engine='bash', eval=FALSE}
project_dir="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10"
region_dir="${project_dir}/project_notes/region_profiles"
cd "$region_dir"

diffbind_dir="${region_dir}/diffbind_sheets"; diffbind_dir="$(readlink -f "$diffbind_dir")"
bigwig_dir="${region_dir}/bigwigs"; bigwig_dir="$(readlink -f "$bigwig_dir")"

bigwig_files="$(find "$bigwig_dir" -type l -name "*.bw")"

bigwig_diffbind_samplesheet="${region_dir}/bigwig_diffbind_samplesheet.tsv"
# clear the sheet
echo -n "" > "$bigwig_diffbind_samplesheet"


# iterate over the bigwig files
for bigwig in $bigwig_files; do
echo "$bigwig"

# parse the file name
bigwig_basename="$(basename "$bigwig")"
echo "$bigwig_basename"
sample_ID="$(echo "$bigwig_basename" | cut -d '-' -f1)"
echo "$sample_ID"
status_ID="$(echo "$bigwig_basename" | cut -d '-' -f2)"
echo "$status_ID"
mark_ID="$(echo "$bigwig_basename" | cut -d '-' -f3 | cut -d '.' -f1)"
echo "$mark_ID"

# get the DiffBind file
sample_diffbind_file="$(find "$diffbind_dir" -name "*${mark_ID}*")"

# write to the sample sheet
echo -e "${bigwig}\t${sample_diffbind_file}\t${sample_ID}\t${status_ID}\t${mark_ID}" >> "$bigwig_diffbind_samplesheet"
done

```


### Find Up/Down Regulated Genes, Top/Bottom Expressed Genes

Gene expression data from a microarray for each sample will be used to determine which genes were up or down regulated in each sample. First, we will read in the gene expression data with R, then for each sample a text file will be written with the genes that have increased (`up`) or decreased (`dn`) gene expression. A gene expression log ratio cutoff of 1.5x increase or decrease will be applied; this translates to a value of > 0.58 or < -0.58 ( log2(1.5) = 0.58 ). 

Additionally, the list of genes used will be filtered to only include genes that overlap peaks which were differentially bound by the H3K27AC histone mark in the ChIP-Seq data produced by DiffBind. 

```{r, eval=FALSE}
# project_dir <- "/Users/kellys04/projects/tmp_smithlab"
# region_dir <- file.path(project_dir,"region_profiles")

project_dir <- "/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10"
region_dir <- file.path(project_dir,"project_notes/region_profiles")
setwd(region_dir)

# output location for gene lists
top_bottom_gene_list_dir <- file.path(region_dir,"top_bottom_expressed_genes/gene_lists")

# load the gene expression data
gene_expr_file <-"gene_expression_microarray/GeneExpression18Patients_includes_SRR.sheet_1.tsv"
gene_expr_df <- read.table(gene_expr_file,sep = '\t',header = TRUE, 
                           row.names = 1, quote = "")
gene_expr_samples <- setNames(object = colnames(gene_expr_df), nm = gsub(pattern = '.', replacement = '-', x = gsub(pattern = 'Exp.', replacement = '', x = colnames(gene_expr_df)), fixed = TRUE))

# get the top and bottom genes for each sample
for(i in seq_along(gene_expr_samples)){
    gene_sample <- gene_expr_samples[i]
    print(gene_sample)
    top_genes <- rownames(head(gene_expr_df[order(gene_expr_df[[gene_sample]], decreasing = TRUE),], 5000))
    bottom_genes <- rownames(tail(gene_expr_df[order(gene_expr_df[[gene_sample]], decreasing = TRUE),], 5000))
    
    top_file <- file.path(top_bottom_gene_list_dir, paste0(names(gene_sample), '_Top_expressed_genes_list.txt'))
    bottom_file <- file.path(top_bottom_gene_list_dir, paste0(names(gene_sample), '_Bottom_expressed_genes_list.txt'))

    writeLines(text = sapply(X = top_genes, FUN = function(x) paste0('gene_name "', x, '"')), con = top_file)
    writeLines(text = sapply(X = bottom_genes, FUN = function(x) paste0('gene_name "', x, '"')), con = bottom_file)
}


```

In bash, we will use the text files generated to create subsets of the Gencode BED file which only contain the gene ID's that match the genes from the microarray data which had increased or decreased expression. Since this operation takes a decent amount of compute power, we will submit each one to a job on the HPC cluster

```{r, engine='bash', eval=FALSE}
project_dir="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10"
region_dir="${project_dir}/project_notes/region_profiles"
log_dir="${region_dir}/logs"; mkdir -p "$log_dir"
cd "$region_dir"

# gencode genes offical list of all genes
gen_bed="${region_dir}/gencode.v19.annotation.genes_only.bed"

# dir with the gene list patterns to search with
top_bottom_gene_listdir="${region_dir}/top_bottom_expressed_genes/gene_lists"

# dir to put the output filtered beds
top_bottom_gene_bed_dir="${region_dir}/top_bottom_expressed_genes/beds"

mkdir -p "$top_bottom_gene_bed_dir"


job_threads="1"
job_mem="5G"

# filter the top and bottom genes
for patternfile in ${top_bottom_gene_listdir}/*.txt; do
echo "$patternfile"
sampleID="$(basename "$patternfile")"
sampleID="${sampleID%%_list.txt}"
job_name="$sampleID"
echo "$sampleID"

outfile="${top_bottom_gene_bed_dir}/${sampleID}.bed"
echo "$outfile"

qsub -wd "$region_dir" -o :${log_dir}/ -e :${log_dir}/ -pe threaded "$job_threads" -l mem_free="$job_mem" -l h_vmem="$job_mem" -l mem_token="$job_mem" -N "$job_name" <<E0F
set -x
grep -f "$patternfile" "$gen_bed" | cut -f 1-3 > "$outfile"
E0F


echo ""
done


```

```

### deepTools: Compute Matrix

#### Make deepTools sample sheet

Need to make a sample sheet that links each sample with its corresponding Up, Down, and Top, Bottom BED region files to be run. 

```{r, engine='bash', eval=FALSE}
project_dir="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10"
region_dir="${project_dir}/project_notes/region_profiles"
cd "$region_dir"

# dir with the filtered beds to be matched
top_bottom_gene_bed_dir="${region_dir}/top_bottom_expressed_genes/beds"

# dir with our bigwig files
bigwig_dir="${region_dir}/bigwigs"; bigwig_dir="$(readlink -f "$bigwig_dir")"
bigwig_files="$(find "$bigwig_dir" -type l -name "*.bw")"




bigwig_BED_samplesheet="${region_dir}/bigwig_BED_samplesheet.tsv"
# clear the sheet
echo -n "" > "$bigwig_BED_samplesheet"


# iterate over the bigwig files
for bigwig in $bigwig_files; do
echo "$bigwig"

# parse the file name
bigwig_basename="$(basename "$bigwig")"
echo "$bigwig_basename"

full_sampleID="${bigwig_basename//.bw}"
sample_ID="$(echo "$bigwig_basename" | cut -d '-' -f1)"
echo "$sample_ID"
status_ID="$(echo "$bigwig_basename" | cut -d '-' -f2)"
echo "$status_ID"
mark_ID="$(echo "$bigwig_basename" | cut -d '-' -f3 | cut -d '.' -f1)"
echo "$mark_ID"

sample_status_ID="${sample_ID}-${status_ID}"

echo "$full_sampleID"
echo "$sample_status_ID"


# !! DONT PROCESS INPUT SAMPLES
# set -x
case "$mark_ID" in 
  *INPUT*)
    echo "Its an INPUT, skipping"
    continue
    ;;
esac
# set +x

# get the top_bottom_gene_bed_dir file
top_gene_bed_file="$(find "$top_bottom_gene_bed_dir" -name "*${sample_status_ID}_Top_expressed_genes.bed")"
bottom_gene_bed_file="$(find "$top_bottom_gene_bed_dir" -name "*${sample_status_ID}_Bottom_expressed_genes.bed")"

# ex:
# SRR-R_Down_regulated_genes.bed
# ZNK-D_Bottom_expressed_genes.bed

echo "$top_gene_bed_file"
echo "$bottom_gene_bed_file"


# write to the sample sheet
echo -e "${bigwig}\t${top_gene_bed_file}\t${bottom_gene_bed_file}\t${up_gene_bed_file}\t${down_gene_bed_file}\t${sample_ID}\t${status_ID}\t${mark_ID}\t${full_sampleID}" >> "$bigwig_BED_samplesheet"

echo ""
echo "-------------------"
done



```

#### Run computeMatrix

Run each D and R sample against its own up & down regulated genes, and top and bottom expressed genes. 

```{r, engine='bash', eval=FALSE}
# run DeepTools make matricies
project_dir="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10"
region_dir="${project_dir}/project_notes/region_profiles"
cd "$region_dir"

bigwig_BED_samplesheet="${region_dir}/bigwig_BED_samplesheet.tsv"

# matrix params
bin_size="50"
region_size="5000"

bigwig_dir="${region_dir}/bigwigs"; bigwig_dir="$(readlink -f "$bigwig_dir")"
matrix_dir="${region_dir}/deepTools_profiles_per_group-bin${bin_size}bp"
mkdir -p "$matrix_dir"
matrix_dir="$(readlink -f "$matrix_dir")"


# qsub params
job_threads="8"
job_mem="4G"
job_options="-j y" # merge stderr and stdout # job_options="-l mem_free=$job_mem -l h_vmem=$job_mem -l mem_token=$job_mem"

# computeMatrix settings
compute_settings='--referencePoint TSS  --averageTypeBins sum '


# iterate over lines in the sample sheet
while read line; do
set -x
if [ ! -z "$line" ]; then

# parse line
sample_bigwig="$(echo "$line" | cut -f1)"
echo "$sample_bigwig"
sample_ID="$(echo "$line" | cut -f6)"
status_ID="$(echo "$line" | cut -f7)"
mark_ID="$(echo "$line" | cut -f8)"
full_sampleID="$(echo "$line" | cut -f9)"

# qsub job name
job_name="$full_sampleID"

# make a list of the BED files to use
# sample_Up_Down_bed_files="$(echo "$line" | cut -f4-5 | tr '\t' ' ')"
sample_Top_Bottom_bed_files="$(echo "$line" | cut -f2-3 | tr '\t' ' ')"
echo "$sample_Up_Down_bed_files"
echo "$sample_Top_Bottom_bed_files"

# make the output labels to use for each
sample_Top_Bottom_label="${full_sampleID}_Top_Bottom_expressed_genes"

sample_Top_Bottom_matrix_file="${matrix_dir}/${sample_Top_Bottom_label}_matrix.gz"
sample_Top_Bottom_heatmap_file="${matrix_dir}/${sample_Top_Bottom_label}_heatmap.png"
sample_Top_Bottom_profile_file="${matrix_dir}/${sample_Top_Bottom_label}_profile.png"
sample_Top_Bottom_logdir="${matrix_dir}/${sample_Top_Bottom_label}_logs"; mkdir -p "$sample_Top_Bottom_logdir"

## # # # ## ## # 
# submit the Top Bottom job
## # # # ## ## # 
qsub -wd "$matrix_dir" -o :${sample_Top_Bottom_logdir}/ -e :${sample_Top_Bottom_logdir}/ -pe threaded "$job_threads" -N "$job_name" $job_options <<E0F
module unload python
unset PYTHONPATH
set -x

cd "$region_dir"
source venv/bin/activate

cd "$matrix_dir"

# make the profile matrix
computeMatrix reference-point --scoreFileName $sample_bigwig --regionsFileName $sample_Top_Bottom_bed_files  -a $region_size -b $region_size --outFileName "$sample_Top_Bottom_matrix_file" --numberOfProcessors "$job_threads" --binSize $bin_size $compute_settings

plotProfile -m "$sample_Top_Bottom_matrix_file" -out "$sample_Top_Bottom_profile_file" --plotTitle "${full_sampleID}"  --perGroup # --numPlotsPerRow 2 # --regionsLabel sample_beds_labels --samplesLabel $bigwig_files_labels

plotProfile -m "$sample_Top_Bottom_matrix_file" --kmeans 2 --plotType heatmap -out "$sample_Top_Bottom_heatmap_file" --perGroup 

E0F




echo ""
fi
set +x
done < "$bigwig_BED_samplesheet"


```

#### Custom Profile Plots

Use R to create custom profile plots for each matrix

```

```{r, engine='bash', eval=FALSE}
project_dir="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10"
region_dir="${project_dir}/project_notes/region_profiles"
cd "$region_dir"

# matrix_dir="${region_dir}/deepTools_profiles_per_group-bin${bin_size}bp"
matrix_dir="${region_dir}/deepTools_profiles_per_group-bin50bp"
plot_output_dir="${region_dir}/ggplot_profiles"

matrix_files="$(find "$matrix_dir" -name "*matrix.gz" -name "*Top_Bottom*")"

for mat_file in $matrix_files; do
echo "$mat_file"
file_basename="$(basename "$mat_file")"
output_name="${file_basename//matrix.gz}ggplot_profile.pdf"
output_path="${plot_output_dir}/${output_name}"
output_logdir="${plot_output_dir}/${output_name//.pdf}/logs"
mkdir -p "$output_logdir"
echo "$output_path"


qsub -wd $PWD -o :${output_logdir}/ -j y <<E0F
pwd
Rscript - "${mat_file}" "${output_path}" <<E0F2
args <- commandArgs(TRUE)

input_file <- args[1]
output_file <- args[2]

cat("input file is:\n")
input_file

cat("output file is:\n")
output_file
sessionInfo()

region_profile_plot <- function(matrix_file_gz, plot_pdf_output_file){
    region_df <- read.table(gzfile(matrix_file_gz), header=FALSE, na.strings = "nan", comment.char = "@")
    # groups = BED region sets = rows
    # samples = BigWig sample files = columns
    
    gz <- gzfile(matrix_file_gz, open = "rt"); sample_JSON <- readLines(gz, 1); close(gz)
    sample_JSON <- gsub(pattern = '^@', replacement = '', x = sample_JSON)
    
    # install.packages("RJSONIO")
    library("RJSONIO")
    param_list <- fromJSON(content = sample_JSON, asText = TRUE)
    
    # remove the chom cols
    chrom_df <- region_df[1:6]
    region_df <- region_df[7:ncol(region_df)]
    
    sample1_df <- region_df[1:param_list[["sample_boundaries"]][2]]
    
    
    sample1_profile <- colSums(region_df[1:param_list[["sample_boundaries"]][2]], na.rm=TRUE)
    
    
    library("ggplot2")
    
    # empty df to hold the data
    profile_df <- data.frame()
    
    # iterate over samples 
    for(i in seq_along(param_list[["sample_labels"]])){ 
        print("i is:")
        print(i)
        
        print("Sample is:")
        sampleID <- param_list[["sample_labels"]][i]
        print(sampleID)
        
        print("Sample Start:")
        sample_start <- param_list[["sample_boundaries"]][i]
        sample_start <- as.numeric(sample_start) + 1
        print(sample_start)
        
        print("Sample End:")
        sample_end <- param_list[["sample_boundaries"]][i + 1]
        print(sample_end)
        
        # iterate over groups 
        for(q in seq_along(param_list[["group_labels"]])){
            print("q is:")
            print(q)
            
            print("Group is:")
            groupID <- param_list[["group_labels"]][q]
            print(groupID)
            
            print("Group start:")
            group_start <- param_list[["group_boundaries"]][q]
            group_start <- as.numeric(group_start) + 1
            print(group_start)
            
            print("Group End:")
            group_end <- param_list[["group_boundaries"]][q + 1]
            print(group_end)
            
            # samples = columns = BigWigs, groups = rows = BEDs
            
            profile_vec <- colSums(region_df[group_start:group_end, sample_start:sample_end], na.rm=TRUE)
            
            # normalize profile values by number of genes
            profile_vec <- profile_vec /(group_end - group_start)
            
            # bin regions for the plot
            bin_regions <- seq(from = (0 - param_list[["downstream"]]), to = param_list[["upstream"]], by = param_list[["bin size"]]) # seq_along(profile_vec)
            # remove '0' bin because we're off by 1... ???
            bin_regions <- bin_regions[! bin_regions %in% 0]
            
            sample_profile_df <- data.frame(profile=profile_vec,
                                            sample=rep(x = sampleID, times = length(profile_vec)),
                                            group=rep(x = groupID, times = length(profile_vec)),
                                            distance_from_TSS= bin_regions 
                                            )
            
            profile_df <- rbind(profile_df, sample_profile_df)
            print(head(sample_profile_df))
            
            cat("\n")
        }# iterate over groups 
    } # iterate over samples 
    
    
    # all four on one graph
    # ggplot(data = profile_df, aes(x=entry, y=profile, shape=group, col=interaction(sample,group))) + geom_line(aes(group=interaction(sample,group)))
    
    
    # separate per sample D R
    pdf(file = plot_pdf_output_file, width = 8, height = 8)
    theme_set(theme_bw())
    print(ggplot(data = profile_df, aes(x=distance_from_TSS, y=profile, col=group)) + geom_line() + facet_wrap(~sample))
    dev.off()
    
    # make the data global
    profile_df <<- profile_df
    param_list <<- param_list
    seq()
} # end of function


# test
# project_dir <- "/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10/project_notes/region_profiles"
# outdir <- file.path(project_dir, "ggplot_profiles")
# matrix_dir <- file.path(project_dir, "deepTools_profiles_per_group-bin50bp")
# matrix_files <- dir(matrix_dir, pattern = "matrix.gz", full.names = TRUE)
# matrix_files <- setNames(object = matrix_files, nm = file.path(outdir, paste0(gsub(pattern = '_matrix.gz', replacement = '', x = basename(matrix_files)), '_ggplot_profile.pdf')))
# region_profile_plot(matrix_file_gz = matrix_files[1], plot_pdf_output_file = names(matrix_files[1]))


save.image(file = file.path(dirname(output_file), "start_session.Rdata"), compress = TRUE, safe = FALSE)
save.image(file = "start_session.Rdata", compress = TRUE, safe = FALSE)

# make region_profile_plot
region_profile_plot(matrix_file_gz = input_file, plot_pdf_output_file = output_file)

save.image(file = "final_start_session.Rdata", compress = TRUE, safe = FALSE)
save.image(file = file.path(dirname(output_file), "final_start_session.Rdata"), compress = TRUE, safe = FALSE)

E0F2
E0F


done





```

convert all the individual PDFs to a multipage PDF

```{r, engine='bash', eval=FALSE}
project_dir="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10"
region_dir="${project_dir}/project_notes/region_profiles"
cd "$region_dir"

plot_output_dir="${region_dir}/ggplot_profiles"
merged_plot_file="${region_dir}/ggplot_region_profiles_merged-2.pdf"
FILES="$(find "$plot_output_dir" -type f -name "*profile.pdf" | sort)"
gs -dBATCH -dNOPAUSE -q -sDEVICE=pdfwrite -sOutputFile="$merged_plot_file" $FILES
```

#### Averaged Profile Plots

Now use the matricies to create a single plot with the average D vs. R values for the Top/Bottom genes

```

```{r, eval=FALSE}
# # # # # # # # # # # # # # 
# workflow functions
# # # # # # # # # # # # # # 
get_matrix_files <- function(status_mark, profile_matrix_dir, ...){
    # get the matrix gz files from the dir with the correct mark in the name
    file_names <- grep(pattern = status_mark, 
                       x = grep(pattern = "Top_Bottom", 
                                x = dir(path = profile_matrix_dir, 
                                        pattern = ".gz", ...), 
                                fixed = TRUE, 
                                value = TRUE), 
                       value = TRUE)
    return(file_names)
}

# make a list of all D and R per mark
make_DR_matrix_files_vec <- function(status_mark, profile_matrix_dir){
    files_vec <- setNames(object = get_matrix_files(status_mark = status_mark, 
                                                    profile_matrix_dir = profile_matrix_dir, 
                                                    full.names = TRUE),
                          nm = get_matrix_files(status_mark = status_mark, 
                                                profile_matrix_dir = profile_matrix_dir))
    return(files_vec)
}

make_DR_matrix_files_list <- function(status_marks_vec, profile_matrix_dir){
    # make a files vec for all entries in the vect
    DR_matrix_files_list <- sapply(status_marks_vec, simplify = FALSE, USE.NAMES = TRUE, function(x) make_DR_matrix_files_vec(status_mark = x, profile_matrix_dir))
    return(DR_matrix_files_list)
}


samples_profile_df <- function(matrix_file_gz){
    # make the dataframe for a sample
    region_df <- read.table(gzfile(matrix_file_gz), header=FALSE, na.strings = "nan", comment.char = "@")
    # groups = BED region sets = rows
    # samples = BigWig sample files = columns
    
    gz <- gzfile(matrix_file_gz, open = "rt"); sample_JSON <- readLines(gz, 1); close(gz)
    sample_JSON <- gsub(pattern = '^@', replacement = '', x = sample_JSON)
    
    # install.packages("RJSONIO")
    library("RJSONIO")
    param_list <- fromJSON(content = sample_JSON, asText = TRUE)
    
    # remove the chom cols
    chrom_df <- region_df[1:6]
    region_df <- region_df[7:ncol(region_df)]
    
    # sample1_df <- region_df[1:param_list[["sample_boundaries"]][2]]
    # sample1_profile <- colSums(region_df[1:param_list[["sample_boundaries"]][2]], na.rm=TRUE)
    
    library("ggplot2")
    
    # empty df to hold the data
    profile_df <- data.frame()
    profile_vec_df <- data.frame()
    
    # iterate over samples 
    for(i in seq_along(param_list[["sample_labels"]])){ 
        print("i is:")
        print(i)
        
        print("Sample is:")
        sampleID <- param_list[["sample_labels"]][i]
        print(sampleID)
        
        print("Sample Start:")
        sample_start <- param_list[["sample_boundaries"]][i]
        sample_start <- as.numeric(sample_start) + 1
        print(sample_start)
        
        print("Sample End:")
        sample_end <- param_list[["sample_boundaries"]][i + 1]
        print(sample_end)
        
        # iterate over groups 
        for(q in seq_along(param_list[["group_labels"]])){
            print("q is:")
            print(q)
            
            print("Group is:")
            groupID <- param_list[["group_labels"]][q]
            print(groupID)
            
            print("Group start:")
            group_start <- param_list[["group_boundaries"]][q]
            group_start <- as.numeric(group_start) + 1
            print(group_start)
            
            print("Group End:")
            group_end <- param_list[["group_boundaries"]][q + 1]
            print(group_end)
            
            # samples = columns = BigWigs, groups = rows = BEDs
            
            profile_vec <- colSums(region_df[group_start:group_end, sample_start:sample_end], na.rm=TRUE)
            
            # normalize profile values by number of genes
            profile_vec <- profile_vec /(group_end - group_start)
            
            # bin regions for the plot
            bin_regions <- seq(from = (0 - param_list[["downstream"]]),
                               to = param_list[["upstream"]],
                               by = param_list[["bin size"]]) # seq_along(profile_vec)
            # remove '0' bin because we're off by 1... ???
            bin_regions <- bin_regions[! bin_regions %in% 0]
            
            sample_profile_df <- data.frame(profile=profile_vec,
                                            sample=rep(x = sampleID, times = length(profile_vec)),
                                            group=rep(x = groupID, times = length(profile_vec)),
                                            distance_from_TSS= bin_regions)
            
            profile_df <- rbind(profile_df, sample_profile_df)
            # profile_vec_df
            profile_vec_df <- rbind(profile_vec_df, profile_vec)
            # print(head(sample_profile_df))
            
            cat("\n")
        }# iterate over groups 
    } # iterate over samples 
    return(profile_df)
} 


plot_profiles <- function(profile_df, plot_pdf_output_file = FALSE, width = 8, height = 8){
    # separate per sample D R
    if(plot_pdf_output_file != FALSE) pdf(file = plot_pdf_output_file, 
                                          width = width, 
                                          height = height)
    theme_set(theme_bw())
    print(ggplot(data = profile_df, 
                 aes(x=distance_from_TSS, 
                     y=profile, 
                     col=group)) + geom_line() + facet_wrap(~sample))
    if(plot_pdf_output_file != FALSE) dev.off()
    
    # all four on one graph
    # ggplot(data = profile_df, 
    #        aes(x=entry, 
    #            y=profile, 
    #            shape=group, 
    #            col=interaction(sample,group))) + geom_line(aes(group=interaction(sample,group)))
}


make_df_per_mark <- function(matrix_files_vec){
    # make a list of sample df's for all entries 
    mark_df_list <- sapply(matrix_files_vec, samples_profile_df, simplify = FALSE, USE.NAMES = TRUE)
    return(mark_df_list)
}


make_full_df <- function(status_mark_list, outdir){
    # start an empty dataframe to hold concatenated df's from each sample
    full_df <- data.frame()
    for(i in seq_along(status_mark_list)){
        # iterate over D and R lists
        group_name <- names(status_mark_list)[i]
        print(group_name)
        # start an empty df to hold the concatenated df's for the group
        group_df <- data.frame()
        for(q in seq_along(status_mark_list[[group_name]])){
            # iterate over the samples in the group
            item_df_name <- names(status_mark_list[[group_name]][q])
            # add the sample's df to the group df
            group_df <- rbind(group_df, status_mark_list[[group_name]][[item_df_name]])
        }
        # add the group df to the full df across all groups
        full_df <- rbind(full_df, group_df)
    }
    
    # output dir
    
    
    
    # make plot of all individual samples
    plot_profiles(profile_df = full_df, plot_pdf_output_file = file.path(outdir, paste0(group_name, "_region_plot_per_sample.pdf")), width = 18, height = 18)
    
    
    
    # split the group column into D and R, Top and Bottom
    groupings_df <- as.data.frame(do.call(rbind, strsplit(x = as.character(full_df[["group"]]), split = '_', perl = TRUE)))
    groupings_df <- cbind(groupings_df, 
                          as.data.frame(do.call(rbind, 
                                                strsplit(as.character(groupings_df$V1), '-'))))
    colnames(groupings_df) <- c("sample_status", "gene_expression", 
                                "expressed", "genes_bed", "sample", 
                                "status")
    groupings_df <- groupings_df[, colnames(groupings_df)[! colnames(groupings_df) %in% c("expressed", "genes_bed")] ]
    
    # add the grouping information to the full df and remove the old cols
    full_df <- full_df[, colnames(full_df)[! colnames(full_df) %in% c("sample", "group")]]
    full_df <- cbind(full_df, groupings_df)
    
    # generate a df with just the avg value for each profile region across samples
    aggr_df <- aggregate(profile ~ gene_expression + status + distance_from_TSS, data = full_df, FUN = mean)
    head(aggr_df)
    
    
    
    # plot average value profiles
    pdf(file = file.path(outdir, paste0(group_name, "_region_plot.pdf")), width = 8, height = 8)
    print(ggplot(data = aggr_df, aes(x=distance_from_TSS, 
                                     y=profile, 
                                     col=gene_expression)) + geom_line() + facet_wrap(~status))
    dev.off()
    
    return(aggr_df)
}

custom_profile_workflow <- function(project_dir, profile_matrix_dir, status_mark_key_df, outdir){
    
    # make a list out of the samplesheet
    mark_list <- setNames(split(t(status_mark_key_df), seq(nrow(t(status_mark_key_df)))), 
                          rownames(t(status_mark_key_df)))
    names(mark_list)
    str(mark_list)
    
    # make a list of all the matrix files
    mark_files_list <- sapply(mark_list, function(x) make_DR_matrix_files_list(status_marks_vec = x, profile_matrix_dir = profile_matrix_dir), simplify = FALSE, USE.NAMES = TRUE)
    
    str(mark_files_list)
    str(mark_files_list[[1]])
    names(mark_files_list)
    names(mark_files_list[[1]][[1]])
    mark_files_list[[1]][[1]]
    
    
    # make a list to hold all the sample df's for all marks
    tmp3 <- sapply(mark_files_list, simplify = FALSE, USE.NAMES = TRUE, function(x) sapply(x, make_df_per_mark, simplify = FALSE, USE.NAMES = TRUE))
    str(tmp3)
    
    # run the full df thing for all samples
    end_list <- sapply(tmp3, simplify = FALSE, USE.NAMES = TRUE, function(x) make_full_df(x, outdir))
    return(end_list)
}

# # # # # # # # # # # # # # 
# # # # # # # # # # # # # # 
# # # # # # # # # # # # # # 

# # # # # # # # # # # # # # 
# workflow
# # # # # # # # # # # # # # 



# locations of the files
project_dir <- "/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10/project_notes/region_profiles"
profile_matrix_dir <- file.path(project_dir, "deepTools_profiles_per_group-bin50bp")
outdir <- file.path(project_dir, "ggplot_custom_profiles")
# sample sheet
status_mark_key <- "H3K27AC\tH3K27ME3\tH3K4ME3\tH3K9AC\tH3K9ME3
D-H3K27AC\tD-H3K27ME3\tD-H3K4ME3\tD-H3K9AC\tD-H3K9ME3
R-H3K27AC\tR-H3K27ME3\tR-H3K4ME3\tR-H3K9AC\tR-H3K9ME3"
status_mark_key_df <- read.table(textConnection(status_mark_key), header = TRUE, sep = '\t')


# run full workflow
end_list <- custom_profile_workflow(project_dir, profile_matrix_dir, status_mark_key_df, outdir)

# make a single plot with all aggregated data
str(end_list)
head(end_list[[1]])
full_aggr_df <- data.frame()
for(i in seq_along(end_list)){
    list_df <- end_list[[i]]
    mark <- names(end_list[i])
    list_df['mark'] <- mark
    full_aggr_df <- rbind(full_aggr_df, list_df)
}

head(full_aggr_df)
# plot average value profiles
pdf(file = file.path(outdir, "all_marks_region_plot.pdf"), width = 12, height = 12)
print(ggplot(data = full_aggr_df, aes(x=distance_from_TSS, 
                                 y=profile, 
                                 col=gene_expression)) + geom_line() + facet_wrap(~mark*status))
dev.off()

```

#### Aris' Averaged Profile Plots

Try re-generating the profile data with Aris' gtools method; code:

```
# tcsh
set genes = diff_gene_beds/AGK_up_genes_Gencode.bed
set ref = ref.bed
cat $genes | gtools-regions pos -op 5p | gtools-regions shiftp -5p -5000 -3p +5000 >! $ref            # flank TSS by 5kb
set bams = alignments/AGK-D-H3K27AC/alignments.bam
set out = test.tsv

gtools-threaded matrix -v -i -o $out -i -nbins 100 -rpkm -profile sum $bams $ref
```

Need to use BAMs for this; make a samplesheet that includes them...

```{r, engine='bash', eval=FALSE}
project_dir="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10"
region_dir="${project_dir}/project_notes/region_profiles"
cd "$region_dir"

# matrix_dir="${region_dir}/deepTools_profiles_per_group-bin${bin_size}bp"
gtools_matrix_dir="${region_dir}/gtools-matrix-ouptut"
plot_output_dir="${gtools_matrix_dir}/ggplot_profiles"
mkdir -p "$plot_output_dir"

# path to the pipeline outputs
align_results_dir="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10/pipeline/align/results" # match full sampleID
diffbind_results_dir="/ifs/data/sequence/results/carrolllab/2016-01-04/ChIP-Seq/diffbind_DESeq2/" # match sample mark
top_bottom_genes_dir="${region_dir}/top_bottom_expressed_genes/beds"
up_down_genes_dir="${region_dir}/up_down_regulated_genes/beds"
samplesheet_in="${region_dir}/samples_samplesheet.tsv"
samplesheet_out="${region_dir}/BAM_bigwig_BED_samplesheet.tsv"

# iterate over the sample IDs in the sheet
echo -n '' > "$samplesheet_out"
while read line; do
if [ ! -z "$line" ]; then
echo "$line"
sample_patient_status_mark="$(echo "$line" | cut -f1)" # ZGR-D-INPUT
sample_mark="$(echo "$line" | cut -f2)"  # INPUT
sample_patient="$(echo "$line" | cut -f3)" # ZGR
sample_patient_mark="$(echo "$line" | cut -f4)" # ZGR-INPUT
sample_status="$(echo "$line" | cut -f5)" # D

sample_bam="$(find "${align_results_dir}" -name "*.bam" -path "*${sample_patient_status_mark}*" )"
sample_bigwig="$(find "${align_results_dir}" -name "*.bw" -path "*${sample_patient_status_mark}*" )"
sample_diffbind="$(find "${diffbind_results_dir}" -name "*.csv" -name "diff_bind.*" -path "*${sample_mark}*")"
sample_top_genes="$(find "${top_bottom_genes_dir}" -name "*${sample_patient_status_mark}*" -name "*Top_expressed_genes.bed" )"
sample_bottom_genes="$(find "${top_bottom_genes_dir}" -name "*${sample_patient_status_mark}*" -name "*Bottom_expressed_genes.bed" )"
sample_up_genes="$(find "${up_down_genes_dir}" -name "*${sample_patient_status_mark}*" -name "*Up_regulated_genes.bed" )"
sample_down_genes="$(find "${up_down_genes_dir}" -name "*${sample_patient_status_mark}*" -name "*Down_regulated_genes.bed" )"


echo "$sample_status"
echo "$sample_bam"
echo "$sample_bigwig"
echo "$sample_diffbind"

echo "$sample_top_genes"
echo "$sample_bottom_genes"
echo "$sample_up_genes"
echo "$sample_down_genes"

# write it to the samplesheet
echo -e "${sample_bam}\t${sample_bigwig}\t${sample_diffbind}\t${sample_top_genes}\t${sample_bottom_genes}\t${sample_up_genes}\t${sample_down_genes}\t${sample_patient_status_mark}\t${sample_mark}\t${sample_patient}\t${sample_patient_mark}\t${sample_status}" >> "$samplesheet_out"
echo ""
fi
done < "$samplesheet_in"



```

run gtools profile command for the items in the sample sheet

```{r, engine='bash', eval=FALSE}
project_dir="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10"
region_dir="${project_dir}/project_notes/region_profiles"
cd "$region_dir"

# matrix_dir="${region_dir}/deepTools_profiles_per_group-bin${bin_size}bp"
gtools_matrix_dir="${region_dir}/gtools-matrix-ouptut"
gtools_extended_regions="${gtools_matrix_dir}/extended_regions"
plot_output_dir="${gtools_matrix_dir}/ggplot_profiles"

mkdir -p "$gtools_extended_regions"
mkdir -p "$plot_output_dir"

samplesheet="${region_dir}/BAM_bigwig_BED_samplesheet.tsv"

# set genes = diff_gene_beds/AGK_up_genes_Gencode.bed
# set ref = ref.bed
# cat $genes | gtools-regions pos -op 5p | gtools-regions shiftp -5p -5000 -3p +5000 >! $ref            # flank TSS by 5kb
# set bams = alignments/AGK-D-H3K27AC/alignments.bam
# set out = test.tsv
# gtools-threaded matrix -v -i -o $out -i -nbins 100 -rpkm -profile sum $bams $ref

# bamfile="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10/pipeline/align/results/align.by_sample.bowtie2/AGK-D-H3K27AC/alignments.bam"
# top_genes="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10/project_notes/region_profiles/top_bottom_expressed_genes/beds/AGK-D-H3K27AC_Top_expressed_genes.bed"
# top_extended_regions="${gtools_extended_regions}/$(basename "$top_genes")_extended.bed"
# top_profile_output="${plot_output_dir}/$(basename "$top_genes")_profile.tsv"

# cat "$top_genes" | gtools-regions pos -op 5p | gtools-regions shiftp -5p -5000 -3p +5000 > "$top_extended_regions"
# gtools-threaded matrix -v -i -o "$top_profile_output" -i -nbins 100 -rpkm -profile sum "$bamfile" "$top_extended_regions"


job_threads="4"
job_mem="10G"

while read line; do
if [ ! -z "$line" ]; then
(
sample_bam="$(echo "$line" | cut -f1)"
sample_top_genes="$(echo "$line" | cut -f4)"
sample_bottom_genes="$(echo "$line" | cut -f5)"

sample_patient_status_mark="$(echo "$line" | cut -f8)" # ZGR-D-INPUT
sample_mark="$(echo "$line" | cut -f9)"  # INPUT
sample_patient="$(echo "$line" | cut -f10)" # ZGR
sample_patient_mark="$(echo "$line" | cut -f11)" # ZGR-INPUT
sample_status="$(echo "$line" | cut -f12)" # D

sample_top_extended_regions="${gtools_extended_regions}/${sample_patient_status_mark}_Top_regions_extended.bed"
sample_bottom_extended_regions="${gtools_extended_regions}/${sample_patient_status_mark}_Bottom_regions_extended.bed"

sample_top_profile_output="${plot_output_dir}/${sample_patient_status_mark}_Top_profile.tsv"
sample_bottom_profile_output="${plot_output_dir}/${sample_patient_status_mark}_Bottom_profile.tsv"

sample_logdir="${gtools_matrix_dir}/logs/${sample_patient_status_mark}"
mkdir -p "$sample_logdir"

echo "$sample_patient_status_mark"
echo "$sample_mark"
echo "$sample_patient"
echo "$sample_patient_mark"
echo "$sample_status"

echo "$sample_bam"
echo "$sample_top_genes"
echo "$sample_bottom_genes"



if [ "INPUT" != "$sample_mark" ]; then

# submit to qsub to run
qsub -wd "$region_dir" -o :${sample_logdir}/ -e :${sample_logdir}/ -j y -pe threaded "$job_threads" -l mem_free="$job_mem" -l h_vmem="$job_mem" -l mem_token="$job_mem" -N "$sample_patient_status_mark" <<E0F
set -x

cat "$sample_top_genes" | gtools-regions pos -op 5p | gtools-regions shiftp -5p -5000 -3p +5000 > "$sample_top_extended_regions"
gtools-threaded matrix -v -i -o "$sample_top_profile_output" -i -nbins 100 -rpkm -profile sum "$sample_bam" "$sample_top_extended_regions"

cat "$sample_bottom_genes" | gtools-regions pos -op 5p | gtools-regions shiftp -5p -5000 -3p +5000 > "$sample_bottom_extended_regions"
gtools-threaded matrix -v -i -o "$sample_bottom_profile_output" -i -nbins 100 -rpkm -profile sum "$sample_bam" "$sample_bottom_extended_regions"

E0F

fi

)


fi
done < "$samplesheet"



```

```

Now plot the profile in R

```{r, eval=FALSE}
library("ggplot2")
profile_dir <- "/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10/project_notes/region_profiles/gtools-matrix-ouptut/ggplot_profiles"

# sample sheet
status_mark_key <- "H3K27AC\tH3K27ME3\tH3K4ME3\tH3K9AC\tH3K9ME3
D-H3K27AC\tD-H3K27ME3\tD-H3K4ME3\tD-H3K9AC\tD-H3K9ME3
R-H3K27AC\tR-H3K27ME3\tR-H3K4ME3\tR-H3K9AC\tR-H3K9ME3"
status_mark_key_df <- read.table(textConnection(status_mark_key), header = TRUE, sep = '\t')



# profile_files <- grep(pattern = as.character(status_mark_key_df[1,1]), x = dir(profile_dir, pattern = "profile.tsv", full.names = TRUE), value = TRUE)
# profile_file <- profile_files[1]

profile_files <- dir(profile_dir, pattern = "profile.tsv", full.names = TRUE)

make_region_df <- function(profile_file){
    # parse the profile file into a dataframe for the sample
    profile_sample_status_mark <- unlist(strsplit(basename(profile_file), '_'))[1]
    profile_sample_ID_vec <- unlist(strsplit(profile_sample_status_mark, '-'))
    profile_status <- profile_sample_ID_vec[2]
    profile_mark <- profile_sample_ID_vec[3]
    profile_sample <- profile_sample_ID_vec[1]
    profile_sample_mark <- sprintf(fmt = '%s-%s', profile_sample_ID_vec[2], profile_sample_ID_vec[3])
    profile_status_mark <- sprintf(fmt = '%s-%s', profile_sample_ID_vec[2], profile_sample_ID_vec[3])
    profile_sample_status <- sprintf(fmt = '%s-%s', profile_sample_ID_vec[1], profile_sample_ID_vec[2])
    profile_genes_type <- unlist(strsplit(basename(profile_file), '_'))[2]
    
    region_df <- read.table(file = profile_file, row.names = 1)
    # rownames(region_df) <- basename(dirname(rownames(region_df)))
    
    region_df <- as.data.frame(t(region_df))
    region_df['sample_status_mark'] <- profile_sample_status_mark
    region_df['status_mark'] <- profile_status_mark
    region_df['genes_type'] <- profile_genes_type
    region_df['status'] <- profile_status
    region_df['mark'] <- profile_mark
    region_df['sample'] <- profile_sample
    region_df['sample_status'] <- profile_sample_status
    colnames(region_df)[1] <- 'profile'
    
    distance_from_TSS <- seq(-5000, 5000, nrow(region_df))
    distance_from_TSS <- distance_from_TSS[! distance_from_TSS %in% 0] 
    
    region_df['distance_from_TSS'] <- distance_from_TSS
    return(region_df)
}

# start empty dataframe to hold all the output.. there's gonna be a lot...
profiles_df <- data.frame()
for(i in seq_along(profile_files)){
    sample_file <- profile_files[i]
    print(sample_file)
    # print(head(make_region_df(sample_file)))
    profiles_df <- rbind(profiles_df, make_region_df(sample_file))
}

colnames(profiles_df)
colnames(status_mark_key_df)


# make plots for each histone mark
for(i in seq(ncol(status_mark_key_df))){
    hist_mark <- colnames(status_mark_key_df)[i]
    D_mark <- status_mark_key_df[1,i]
    R_mark <- status_mark_key_df[2,i]
    print(sprintf("%s %s %s", hist_mark, D_mark, R_mark))
    
    profile_subset <- subset(profiles_df, subset=(mark == hist_mark))
    
    pdf(file = file.path(profile_dir, sprintf("%s_gtools_profile_per_sample.pdf", hist_mark)), width = 12, height = 12)
    print(ggplot(data = profile_subset, aes(x=distance_from_TSS, y=profile, col=genes_type)) + geom_line() + facet_wrap(~sample_status))
    dev.off()
    
    head(profiles_df)
    aggr_df <- aggregate(profile ~ genes_type + distance_from_TSS + mark + status, data = profile_subset, FUN = mean)
    head(aggr_df)
    
    pdf(file = file.path(profile_dir, sprintf("%s_gtools_profile_avg.pdf", hist_mark)), width = 12, height = 12)
    print(ggplot(data = aggr_df, aes(x=distance_from_TSS, y=profile, col=genes_type)) + geom_line() + facet_wrap(~status))
    dev.off()

}












    # theme_set(theme_bw())
    # print(ggplot(data = profile_df, 
    #              aes(x=distance_from_TSS, 
    #                  y=profile, 
    #                  col=group)) + geom_line() + facet_wrap(~sample))
    # if(plot_pdf_output_file != FALSE) dev.off()
    # 
    # # all four on one graph
    # # ggplot(data = profile_df, 
    # #        aes(x=entry, 
    # #            y=profile, 
    # #            shape=group, 
    # #            col=interaction(sample,group))) + geom_line(aes(group=interaction(sample,group)))

```

convert all the individual PDFs to a multipage PDF

```{r, engine='bash', eval=FALSE}
project_dir="/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10"
region_dir="${project_dir}/project_notes/region_profiles"
cd "$region_dir"

plot_output_dir="${region_dir}/gtools-matrix-ouptut/ggplot_profiles"
merged_plot_file="${region_dir}/gtools-matrix-ouptut/ggplot_profiles/ggplot_region_profiles_merged-2.pdf"
# FILES="$(find "$plot_output_dir" -type f -name "*profile.pdf" | sort)"
# gs -dBATCH -dNOPAUSE -q -sDEVICE=pdfwrite -sOutputFile="$merged_plot_file" $FILES
```


# Results {.tabset .tabset-pills}

DeepTools profile plots created using the set of all Gencode genes

```{r, results='asis', eval=TRUE}
profiles_dir <- "/ifs/home/kellys04/projects/SmithLab_ChIpSeq_2016-03-10/project_notes/region_profiles/ggplot_profiles"

profile_plots <- sort(dir(path = profiles_dir, pattern = "profile.png", recursive = TRUE, full.names = TRUE))
profile_plots <- setNames(object = profile_plots, nm = gsub(pattern = '_profile.png', replacement = '', x = basename(profile_plots)))

updn_plot <- profile_plots[which(grepl(pattern = 'updn', x = names(profile_plots)))]
hilo_plot <- profile_plots[which(grepl(pattern = 'hilo', x = names(profile_plots)))]

sample_IDs <- unique(gsub(x = names(profile_plots), pattern = '([[:alpha:]]*)(.*)', replacement = '\\1', perl = TRUE))

# Up Down comparisons
cat(paste0('## Up vs. Down  \n \n')) # {.tabset .tabset-pills}
cat('Genes that were up vs. down regulated in the gene expression data from the microarray. \n \n')

for(ID in sample_IDs){
    sample_plots <- updn_plot[grepl(pattern = ID, x = names(updn_plot))]
    # cat(paste0('### ', ID, ' \n \n'))
    
    for(sample_plot in sample_plots){
        cat(paste0('![](',sample_plot,') \n \n'))
    }
}

# Up Down comparisons
cat(paste0('## Top vs. Bottom  \n \n')) # {.tabset .tabset-pills}
cat('The top 3000 genes vs. the bottom 3000 expressed genes from the microarray. \n \n')

for(ID in sample_IDs){
    sample_plots <- hilo_plot[grepl(pattern = ID, x = names(hilo_plot))]
    # cat(paste0('### ', ID, ' \n \n'))
    
    for(sample_plot in sample_plots){
        cat(paste0('![](',sample_plot,') \n \n'))
    }
}



```


