---
title: "ChIP-Seq Differential Peaks"
author: "Stephen Kelly"
date: "9/13/2016"
output: 
  html_document: 
    fig_height: 8
    fig_width: 8
    keep_md: yes
    number_sections: true
    code_folding: hide
    toc: true
    toc_float: true
citation_package: natbib
bibliography: references.bib
biblio-style: apsr
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# knitr::opts_chunk$set(dev = 'pdf')
```

# DiffBind 

DiffBind [@DiffBind] is used to determine which peaks in a ChIP-Seq experiment are differential bound between sample data sets. For this report, we have subset the standard DiffBind data to plot only the single peak per gene which is closest to the gene's start site (e.g. lowest 'shortestDistance' value).


```{r, echo=TRUE}

# function for formatting text in the report
mycat <- function(text){
    cat(gsub(pattern = "\n", replacement = "  \n", x = text))
}

# function to process the DiffBind data
Diff_process_data <- function(diff_df, signif_p_value=0.05, 
                                                               fold_change_value=1, 
                                                               plot_colors=c("red","gray","orange", "blue", "green"), 
                                                               top_gene_number=10){
    # browser() # use this for debugging
    suppressPackageStartupMessages(library("data.table"))
    # this function makes the custom DiffBind volcano plots
    # only plots protein coding promoter genes from DiffBind output
    # highlights top x fold changed genes
    
    
    # ~~~~~~ PROCESS DATA ~~~~~~~~ # 
    # get the colname order for later
    original_colnames <- colnames(diff_df)
    
    
    # GET THE CLOSEST DIFFBIND PEAK PER GENE 
    # in Diffbind subset, getthe peaks closest to each gene
    # # for each gene name, find the peak with the lowest "shortestDistance" value
    # make it a data.table type
    setDT(diff_df)
    # get min value per factor, also remove duplicates
    diff_df_min <- as.data.frame(diff_df[, .SD[which.min(shortestDistance)], by=external_gene_name])
    # fix the colname order
    diff_df_min <- diff_df_min[,original_colnames]
    
    # get sample ID's from 7th, 8th colnames
    sampleID_1 <- colnames(diff_df_min)[7]
    sampleID_2 <- colnames(diff_df_min)[8]
    
    
    # subset for significant genes
    diff_df_min_signiff <- as.data.frame(subset(diff_df_min, p.value<signif_p_value & abs(Fold)>fold_change_value))
    
    # subset for protein coding genes
    diff_df_min_signiff_protein <- as.data.frame(subset(diff_df_min_signiff, gene_biotype=="protein_coding"))
    
    # subset top up/down protein coding genes
    diff_df_min_signiff_UpFold <- diff_df_min_signiff_protein[with(diff_df_min_signiff_protein, 
                                                                   order(Fold, p.value)), 
                                                              c("Fold","p.value","external_gene_name")][1:top_gene_number,]
    diff_df_min_signiff_DnFold <- diff_df_min_signiff_protein[with(diff_df_min_signiff_protein, 
                                                                   order(-Fold, p.value)), 
                                                              c("Fold","p.value","external_gene_name")][1:top_gene_number,]
    # ~~~~~~~~~~~~~~ # 
    
    
    
    
    diff_df_min <-diff_df_min[c("external_gene_name", "gene_biotype", "Fold", "p.value")]
    diff_df_min["group"] <- "not_signif"
    diff_df_min[which(diff_df_min['p.value'] < signif_p_value & abs(diff_df_min['Fold']) < fold_change_value ),"group"] <- "signif"
    diff_df_min[which(diff_df_min['p.value'] > signif_p_value & abs(diff_df_min['Fold']) > fold_change_value ),"group"] <- "fc"
    diff_df_min[which(diff_df_min['p.value'] < signif_p_value & abs(diff_df_min['Fold']) > fold_change_value ),"group"] <- "signif_fc"
        # browser() # use this for debugging


    
    suppressPackageStartupMessages(library(dplyr))

    
    diff_df_min <- diff_df_min %>%
        arrange(desc(gene_biotype == "protein_coding"), desc(Fold), p.value) %>% 
        mutate(group = ifelse(row_number() < top_gene_number + 1 & abs(Fold) > fold_change_value, "top_signif_fc", group))
    
    diff_df_min <- diff_df_min %>%
        arrange(desc(gene_biotype == "protein_coding"), Fold, p.value) %>% 
        mutate(group = ifelse(row_number() < top_gene_number + 1 & abs(Fold) > fold_change_value, "top_signif_fc", group))
    
    return(diff_df_min)
    
}


# Function for Plot.ly volcano plot
DiffBind_volcano_plotly_top_protein_coding_promoters <- function(diff_df, signif_p_value=0.05, 
                                                               fold_change_value=1, 
                                                               plot_colors=c("red","gray","orange", "blue", "green"), 
                                                               top_gene_number=10, ... ){
    
    
    diff_df_min <- Diff_process_data(diff_df)

    # ~~~~~~ MAKE PLOT ~~~~~~~~ # 
    mycat('### Volcano Plot: Plot.ly\n')
    suppressPackageStartupMessages(library("plotly"))
    
    plot_ly(data = diff_df_min, x = Fold, y = -log10(p.value), text = external_gene_name, mode = "markers", color = as.ordered(group), colors = plot_colors)
    
    # my_plot <- # can also save it to an object to be returned
    # print(my_plot)
    # print(htmltools::tagList(list(as.widget(my_plot))))
    # return(my_plot)
    
    
}


Diff_stats <- function(diff_df, signif_p_value=0.05, fold_change_value=1, plot_colors=c("red","gray","orange", "blue", "green"), top_gene_number=10, ... ){
    
    # process the DiffData
    diff_df_min <- Diff_process_data(diff_df)
    
    # print some gene table stats into the report; this could use optimization
    mycat('\n\n')
    mycat('\n******\n')
    # FULL DATASET STATS
    mycat('### Overall DiffBind Stats\n')
    mycat(paste0('Total number of DiffBind peaks:\n', length(as.data.frame(diff_df)[["external_gene_name"]]), '\n\n'))
    mycat(paste0('Total number of DiffBind genes:\n', length(unique(diff_df_min[["external_gene_name"]])), '\n\n'))
    
    mycat(paste0('Total number positive fold change genes:\n', 
                 length(unique(subset(diff_df_min, Fold > 0 )[["external_gene_name"]])), '\n\n'))
    
    mycat(paste0('Total number negative fold change genes:\n', 
                 length(unique(subset(diff_df_min, Fold < 0 )[["external_gene_name"]])), '\n\n'))
    
    # significant 
    mycat('\n******\n')
    mycat(paste0('Total number of p <' ,signif_p_value,' genes:\n', 
                 length(unique(subset(diff_df_min, p.value<signif_p_value )[["external_gene_name"]])), '\n\n'))
    
    mycat(paste0('Total number of p <' ,signif_p_value,' genes (pos. FC):\n', 
                 length(unique(subset(diff_df_min, p.value<signif_p_value &
                                          Fold > 0)[["external_gene_name"]])), '\n\n'))
    
    mycat(paste0('Total number of p <' ,signif_p_value,' genes (neg. FC):\n', 
                 length(unique(subset(diff_df_min, p.value<signif_p_value &
                                          Fold < 0)[["external_gene_name"]])), '\n\n'))
    
    # fold change
    mycat('\n******\n')
    mycat(paste0("Total number of log2(Fold Change) > ",fold_change_value, ' genes:\n',
                 length(unique(subset(diff_df_min, abs(Fold)>fold_change_value)[["external_gene_name"]])), '\n\n'))
    
    mycat(paste0("Total number of log2(Fold Change) > ",fold_change_value, ' genes (pos. FC):\n',
                 length(unique(subset(diff_df_min, abs(Fold)>fold_change_value & 
                                          Fold > 0)[["external_gene_name"]])), '\n\n'))
    
    mycat(paste0("Total number of log2(Fold Change) > ",fold_change_value, ' genes (neg. FC):\n',
                 length(unique(subset(diff_df_min, abs(Fold)>fold_change_value & 
                                          Fold < 0)[["external_gene_name"]])), '\n\n'))
    
    # siginificant & fold change
    mycat('\n******\n')
    mycat(paste0("Total number of p < ",signif_p_value," & log2(Fold Change) > ",fold_change_value, ' genes:\n',
                 length(unique(subset(diff_df_min, p.value<signif_p_value &
                                          abs(Fold)>fold_change_value)[["external_gene_name"]])), '\n\n'))
    
    mycat(paste0("Total number of p < ",signif_p_value," & log2(Fold Change) > ",fold_change_value, ' genes (pos. FC):\n',
                 length(unique(subset(diff_df_min, p.value<signif_p_value & 
                                          Fold > 0 &
                                          abs(Fold)>fold_change_value)[["external_gene_name"]])), '\n\n'))
    
    mycat(paste0("Total number of p < ",signif_p_value," & log2(Fold Change) > ",fold_change_value, ' genes (neg. FC):\n',
                 length(unique(subset(diff_df_min, p.value<signif_p_value & 
                                          Fold < 0 &
                                          abs(Fold)>fold_change_value)[["external_gene_name"]])), '\n\n'))
    
    
    mycat('\n\n')
    mycat('\n******\n')
    # ONLY PROTEIN CODING GENES
    mycat('### Protein Coding Gene Stats\n') # gene_biotype=="protein_coding"
    mycat(paste0('Total number of DiffBind peaks:\n', 
                 length(subset(as.data.frame(diff_df), gene_biotype=="protein_coding")[["external_gene_name"]]), '\n\n'))
    mycat(paste0('Total number of DiffBind genes:\n', 
                 length(unique(subset(diff_df_min, gene_biotype=="protein_coding")[["external_gene_name"]])), '\n\n'))
    
    mycat(paste0('Total number positive fold change genes:\n', 
                 length(unique(subset(diff_df_min, gene_biotype=="protein_coding" & 
                                          Fold > 0 )[["external_gene_name"]])), '\n\n'))
    
    mycat(paste0('Total number negative fold change genes:\n', 
                 length(unique(subset(diff_df_min, gene_biotype=="protein_coding" &
                                          Fold < 0 )[["external_gene_name"]])), '\n\n'))
    
    # significant 
    mycat('\n******\n')
    mycat(paste0('Total number of p <' ,signif_p_value,' genes:\n', 
                 length(unique(subset(diff_df_min, gene_biotype=="protein_coding" &
                                          p.value<signif_p_value )[["external_gene_name"]])), '\n\n'))
    
    mycat(paste0('Total number of p <' ,signif_p_value,' genes (pos. FC):\n', 
                 length(unique(subset(diff_df_min, gene_biotype=="protein_coding" &
                                          p.value<signif_p_value &
                                          Fold > 0)[["external_gene_name"]])), '\n\n'))
    
    mycat(paste0('Total number of p <' ,signif_p_value,' genes (neg. FC):\n', 
                 length(unique(subset(diff_df_min, gene_biotype=="protein_coding" &
                                          p.value<signif_p_value &
                                          Fold < 0)[["external_gene_name"]])), '\n\n'))
    
    # fold change
    mycat('\n******\n')
    mycat(paste0("Total number of log2(Fold Change) > ",fold_change_value, ' genes:\n',
                 length(unique(subset(diff_df_min, gene_biotype=="protein_coding" &
                                          abs(Fold)>fold_change_value)[["external_gene_name"]])), '\n\n'))
    
    mycat(paste0("Total number of log2(Fold Change) > ",fold_change_value, ' genes (pos. FC):\n',
                 length(unique(subset(diff_df_min, gene_biotype=="protein_coding" &
                                          abs(Fold)>fold_change_value & 
                                          Fold > 0)[["external_gene_name"]])), '\n\n'))
    
    mycat(paste0("Total number of log2(Fold Change) > ",fold_change_value, ' genes (neg. FC):\n',
                 length(unique(subset(diff_df_min, gene_biotype=="protein_coding" &
                                          abs(Fold)>fold_change_value & 
                                          Fold < 0)[["external_gene_name"]])), '\n\n'))
    
    # siginificant & fold change
    mycat('\n******\n')
    mycat(paste0("Total number of p < ",signif_p_value," & log2(Fold Change) > ",fold_change_value, ' genes:\n',
                 length(unique(subset(diff_df_min, gene_biotype=="protein_coding" &
                                          p.value<signif_p_value &
                                          abs(Fold)>fold_change_value)[["external_gene_name"]])), '\n\n'))
    
    mycat(paste0("Total number of p < ",signif_p_value," & log2(Fold Change) > ",fold_change_value, ' genes (pos. FC):\n',
                 length(unique(subset(diff_df_min, gene_biotype=="protein_coding" &
                                          p.value<signif_p_value & 
                                          Fold > 0 &
                                          abs(Fold)>fold_change_value)[["external_gene_name"]])), '\n\n'))
    
    mycat(paste0("Total number of p < ",signif_p_value," & log2(Fold Change) > ",fold_change_value, ' genes (neg. FC):\n',
                 length(unique(subset(diff_df_min, gene_biotype=="protein_coding" &
                                          p.value<signif_p_value & 
                                          Fold < 0 &
                                          abs(Fold)>fold_change_value)[["external_gene_name"]])), '\n\n'))
    
    
    
    mycat('\n')
    mycat('\n******\n')
    mycat('\n******\n')
    
}

# function for the base R plots:
DiffBind_volcano_plot_top_protein_coding_promoters <- function(diff_df, signif_p_value=0.05, 
                                                               fold_change_value=1, 
                                                               plot_colors=c("grey","red","orange","blue", "green"), 
                                                               top_gene_number=10){
    # browser() # use this for debugging
    
    # this function makes the custom DiffBind volcano plots
    # only plots protein coding promoter genes from DiffBind output
    # highlights top x fold changed genes
    
    
    # ~~~~~~ PROCESS DATA ~~~~~~~~ # 
    # get the colname order for later
    original_colnames <- colnames(diff_df)
    
    suppressPackageStartupMessages(library("data.table"))
    # GET THE CLOSEST DIFFBIND PEAK PER GENE 
    # in Diffbind subset, getthe peaks closest to each gene
    # # for each gene name, find the peak with the lowest "shortestDistance" value
    # make it a data.table type
    setDT(diff_df)
    # get min value per factor, also remove duplicates
    diff_df_min <- as.data.frame(diff_df[, .SD[which.min(shortestDistance)], by=external_gene_name])
    # fix the colname order
    diff_df_min <- diff_df_min[,original_colnames]
    
    # get sample ID's from 7th, 8th colnames
    sampleID_1 <- colnames(diff_df_min)[7]
    sampleID_2 <- colnames(diff_df_min)[8]
    
    
    # subset for significant genes
    diff_df_min_signiff <- as.data.frame(subset(diff_df_min, p.value<signif_p_value & abs(Fold)>fold_change_value))
    
    # subset for protein coding genes
    diff_df_min_signiff_protein <- as.data.frame(subset(diff_df_min_signiff, gene_biotype=="protein_coding"))
    
    # subset top up/down protein coding genes
    diff_df_min_signiff_UpFold <- diff_df_min_signiff_protein[with(diff_df_min_signiff_protein, 
                                                                   order(Fold, p.value)), 
                                                              c("Fold","p.value","external_gene_name")][1:top_gene_number,]
    diff_df_min_signiff_DnFold <- diff_df_min_signiff_protein[with(diff_df_min_signiff_protein, 
                                                                   order(-Fold, p.value)), 
                                                              c("Fold","p.value","external_gene_name")][1:top_gene_number,]
    # ~~~~~~~~~~~~~~ # 
    
    # suppressPackageStartupMessages(library("plotly"))
    # browser() # use this for debugging
    
    # ~~~~~~ MAKE PLOT ~~~~~~~~ # 
    mycat('### Volcano Plot: base R\n')
    # set the multi-pane plot layout matrix
    plot_layout_matrix<-structure(c(1L, 2L, 2L, 2L, 2L, 2L, 3L, 3L, 3L, 3L, 
                                    3L, 1L, 2L, 2L, 2L, 2L, 2L, 3L, 3L, 3L, 
                                    3L, 3L, 1L, 2L, 2L, 2L, 2L, 2L, 3L, 3L, 
                                    3L, 3L, 3L, 1L, 2L, 2L, 2L, 2L, 2L, 3L, 
                                    3L, 3L, 3L, 3L), 
                                  .Dim = c(11L,4L), 
                                  .Dimnames = list(NULL, c("V1", "V2", "V3", "V4")))
    
    # subset the matrix for two panels # keep this because I like this code sample
    plot_layout_matrix2 <- plot_layout_matrix[(rowSums(plot_layout_matrix < 3) > 0), , drop = FALSE]
    # x[which(x < 3, arr.ind = TRUE)[,1],]
    
    # SET PLOT LAYOUT
    layout(plot_layout_matrix2) 
    
    
    # FIRST PLOT PANEL
    # adjust first panel margins; these need to be adjusted to intended plot size
    # currently configured for 8 inch x 8 inch plota
    # plot margins # c(bottom, left, top, right) # default is c(5, 4, 4, 2) + 0.1
    par(mar=c(1,3,1,0))
    # call blank plot to fill the first panel
    plot(1,type='n',axes=FALSE,xlab="",ylab="")
    # set up the Legend in the first panel; 1st legend only title, 2nd legend only legend
    legend("center",legend = "",title=paste0("Volcano plot: ", sampleID_1, " vs. ",sampleID_2,"\n"),cex=1.3, bty='n') 
    legend("bottom",legend=c("Not Significant",
                             paste0("p < ",signif_p_value),
                             paste0("log2(Fold Change) > ",fold_change_value),
                             paste0("p < ",signif_p_value," & log2(Fold Change) > ",fold_change_value),
                             paste0("Top log2(Fold Change) genes")),
           fill=plot_colors,bty = "n",ncol=2,cex=0.9)
    
    
    
    # SECOND PLOT PANEL
    # adjust second panel margins
    par(mar=c(6,4,0,3)+ 0.1)
    
    # add base volcano plot; all points
    with(diff_df_min, plot(Fold, -log10(p.value), pch=20, xlim=c(min(Fold)-1,max(Fold)+1),col=plot_colors[1],xlab = "log2(Fold Change)"))
    
    # Add colored points for data subsets
    with(subset(diff_df_min, p.value<signif_p_value ), points(Fold, -log10(p.value), pch=20, col=plot_colors[2]))
    with(subset(diff_df_min, abs(Fold)>fold_change_value), points(Fold, -log10(p.value), pch=20, col=plot_colors[3]))
    with(subset(diff_df_min, p.value<signif_p_value & abs(Fold)>fold_change_value), points(Fold, -log10(p.value), pch=20, col=plot_colors[4]))
    
    # add points and gene labels for top genes
    suppressPackageStartupMessages(library("calibrate"))
    suppressPackageStartupMessages(library("reshape2"))
    # up genes
    with(na.omit(diff_df_min_signiff_UpFold), points(Fold, -log10(p.value), pch=20, col="green"))
    with(na.omit(diff_df_min_signiff_UpFold), textxy(Fold, -log10(p.value), labs=external_gene_name, cex=.8))
    # down genes
    with(na.omit(diff_df_min_signiff_DnFold), points(Fold, -log10(p.value), pch=20, col="green"))
    with(na.omit(diff_df_min_signiff_DnFold), textxy(Fold, -log10(p.value), labs=external_gene_name, cex=.8))
    
    
    
}
```


```{r, results='asis', echo=TRUE}
mycat('# Plots and Results {.tabset}\n\n') # .tabset-fade .tabset-pills

# out_dir <- "/ifs/home/kellys04/projects/Bioinformatics/DiffBind_scripts_reports/DiffBind_Volcano_Plot_report/input"
out_dir <- "/Users/kellys04/Bioinformatics/DiffBind_scripts_reports/DiffBind_Volcano_Plot_report/input"

# mycat(paste0("Project dir:\n",out_dir,'\n'))
# mycat('\n******\n')

diff_cl4_ctrl_file <- paste0(out_dir,"/diff_bind.Treatment4-ChIPSeq-vs-Control-ChIPSeq.p100.csv")
diff_cl4_cl5_file <- paste0(out_dir,"/diff_bind.Treatment4-ChIPSeq-vs-Treatment5-ChIPSeq.p100.csv")
diff_cl5_ctrl_file <- paste0(out_dir,"/diff_bind.Treatment5-ChIPSeq-vs-Control-ChIPSeq.p100.csv")

sample_file_list <- setNames(c(diff_cl4_ctrl_file, diff_cl4_cl5_file, diff_cl5_ctrl_file),
                             c("Sample4_Control", "Sample4_Sample5", "Sample5_Control"))



sample_file <- sample_file_list[1]
mycat(paste0("## ", names(sample_file), ' {.tabset}\n'))
diff_df <- read.delim(file = sample_file,header = TRUE,sep = ',')
DiffBind_volcano_plotly_top_protein_coding_promoters(diff_df = diff_df)
# DiffBind_volcano_plot_top_protein_coding_promoters(diff_df = diff_df)
Diff_stats(diff_df = diff_df)


sample_file <- sample_file_list[2]
mycat(paste0("## ", names(sample_file), ' {.tabset}\n'))
diff_df <- read.delim(file = sample_file,header = TRUE,sep = ',')
DiffBind_volcano_plotly_top_protein_coding_promoters(diff_df = diff_df)
# DiffBind_volcano_plot_top_protein_coding_promoters(diff_df = diff_df)
Diff_stats(diff_df = diff_df)

sample_file <- sample_file_list[3]
mycat(paste0("## ", names(sample_file), ' {.tabset}\n'))
diff_df <- read.delim(file = sample_file,header = TRUE,sep = ',')
DiffBind_volcano_plotly_top_protein_coding_promoters(diff_df = diff_df)
# DiffBind_volcano_plot_top_protein_coding_promoters(diff_df = diff_df)
Diff_stats(diff_df = diff_df)

# LOOP DOES NOT WORK, FIGURE THIS OUT LATER
# DiffBind_volcano_plotly_top_protein_coding_promoters(diff_df = read.delim(file = sample_file_list[1], header = TRUE,sep = ','))

# for(i in seq_along(sample_file_list)){
#     sample_file <- sample_file_list[i]
#     mycat(paste0("## ", names(sample_file), ' {.tabset}\n'))
#     # mycat(paste0("File: ", sample_file, '\n'))
#     diff_df <- read.delim(file = sample_file,header = TRUE,sep = ',')
#     print(DiffBind_volcano_plotly_top_protein_coding_promoters(diff_df = diff_df))
#     DiffBind_volcano_plot_top_protein_coding_promoters(diff_df = diff_df)
#     Diff_stats(diff_df = diff_df)
# }
```

# System Information

```{r}
system('uname -srv',intern=T)
sessionInfo()
# save.image(compress = TRUE, )
```

# References